### DISCUSSION TITLE
lesson 9,  lottery, fulfillRandomWords

### DISCUSSION THREAD

**Discussion by AndroKasrashvili** [Cyfrin/foundry-full-course-cu#1483] [Q&A] [answered]
# Hi there!

### Currently I am working on lottery project, everything goes fine, just I am getting curious, why we are setting raffle state to open before we give away balance to winner?

```solidity
function fulfillRandomWords(
        uint256 /* requestId */,
        uint256[] memory randomWords
    ) internal override {
        // s_players size 10
        // randomNumber 202
        // 202 % 10 ? what's doesn't divide evenly into 202?
        // 20 * 10 = 200
        // 2
        // 202 % 10 = 2
        uint256 indexOfWinner = randomWords[0] % s_players.length;
        address payable recentWinner = s_players[indexOfWinner];
        s_recentWinner = recentWinner;
        s_players = new address payable[](0);
        s_raffleState = RaffleState.OPEN;
        s_lastTimeStamp = block.timestamp;
        emit WinnerPicked(recentWinner);
        (bool success, ) = recentWinner.call{value: address(this).balance}("");
        // require(success, "Transfer failed");
        if (!success) {
            revert Raffle__TransferFailed();
        }
    }

```
I guess, I lack understanding about how atomic it will get executed.

**Comment by EngrPips** [comment]
Hello @AndroKasrashvili , Patrick strived as much as possible to follow `CEI` [checks, effects, interactions]. so if you follow that pattern you want to make sure that you do every effects on your contract before you interact with the outside world so as to prevent reentrancy attack. setting the raffle state to open is us affecting our contract state [that is changing the state of our contract] while sending the balance to the winner is us interacting with the outside world that can actually disrupt our system if we did not take all necessary precautions, The last thing we always want to do in our contract function is interacting with the outside world. hence that is why we reset the state of our contract to open first before actually sending out money to the winner.

I hope i explained well enough for you to understand.

**Reply by AndroKasrashvili** [reply]
Hi, Thanks for reply !

I am just trying to get more context of what's happening behind the scene.

Let's say I have updated Raffle State to open and before winner gets rewarded there is other request to enroll to Raffle and Raffle balance gets increased, so prev winner will get rewarded with money which does not belong to him/her. Is this actual real case? or can we say desperately that until this method as transaction will get mined no one will be able to enroll in Raffle?

Thanks!

**Comment by EngrPips** [comment]
Well transactions don't happen the instant you sent them, so i assume if the `fulfillRandomWords` is running already and then a user sent a transaction to enter the game, the `fulfillRandomWords` function would most likely finish running already before the transaction to enter the game is mined. That's what i assume honestly

**Reply by AndroKasrashvili** [reply]
Thanks for honesty, I do really appreciate that, can we  get rid of assumptions at some point ? 
do you have an idea where it can be researched, my guess is EVM? maybe it's covered in security section? 

Thanks!

**Reply by EngrPips** [reply, marked as answer]
Well, i think this is what you can experiment yourself. i mean you wrote the whole contract and have it working, so you can deploy the lottery contract to sepolia network, generate other address for yourself and fund them with enough sepolia ether then use all of those address to enter the game, the moment you see that chainlink fires the perfomUpkeep call try to enter the lottery with an address and see what the outcome is. it would be a great experiment for you to learn and not just some assumption.
